third_party/nvfuser/runtime/basic_type_traits.cu

```diff
@@ -0,0 +1,172 @@
+namespace std {
```

Comment:
Should we protect this file with
```
#ifdef __NVCC__
#include <type_traits>
#else
....
#endif
```
Please try a `PYTORCH_NVFUSER_DUMP="cuda_to_file"` run on something, and use nvcc to compile the dumped file to check if it compiles.

Reply:
nice catch! we need this protection. However, the generated code still can't compile with nvcc due to the lack of non-volatile to volatile copy of complex numbers in <complex>. error msg:
```
__tmp_kernel1.cu(3426): error: no operator "=" matches these operands
            operand types are: volatile std::complex<float> = const std::complex<float>
          detected during:
            instantiation of "void CudaCodeGen::TupleCopy<DstType, SrcType, num_vals>::copy(DstType &, CudaCodeGen::nvfuser_index_t, const SrcType &, CudaCodeGen::nvfuser_index_t) [with DstType=CudaCodeGen::PtrTupleBase<true, float, double, CudaCodeGen::int64_t, std::complex<float>>, SrcType=CudaCodeGen::LocalTuple<float, double, CudaCodeGen::int64_t, std::complex<float>>, num_vals=4]" 
```

Reply:
Are you saying that, if the kernel is complex, then we are still unable to compile with nvcc. But if it is not complex then it is working? If so, then it is fine for now. I believe @mmigdal-nv is using nvcc with matmul quite often, and we need to make sure this PR doesn't break non-complex matmul use case.

Reply:
It's only influence reduction of complex numbers.